src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -286,52 +439,119 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
     return 1; /* return 1 byte to avoid division by zero */
 }
 
+static ucs_status_t
+uct_cuda_copy_detect_vmm(const void *address, CUmemorytype *cuda_mem_type,
+                         unsigned *is_vmm_allocation)
+{
+    ucs_status_t status = UCS_OK;
+#if HAVE_CUDA_FABRIC
+    CUmemAllocationProp prop = {};
+    CUmemGenericAllocationHandle alloc_handle;
+
+    /* Check if memory is allocated using VMM API and see if host memory needs
+     * to be treated as pinned device memory */
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemRetainAllocationHandle(&alloc_handle, (void*)address));
+    if (status != UCS_OK) {
+        status = UCS_OK;
+        goto ret;
+    }
+
+    *is_vmm_allocation = 1;
+    if (*cuda_mem_type != CU_MEMORYTYPE_HOST) {
+        goto mem_release;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemGetAllocationPropertiesFromHandle(&prop, alloc_handle));
+    if (status != UCS_OK) {
+        status = UCS_ERR_INVALID_ADDR;
+        goto mem_release;
+    }
+
+    /* if location type is HOST the memory can be potentially accessed by CEs */
+    if ((prop.location.type == CU_MEM_LOCATION_TYPE_HOST) ||
+        (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA) ||
+        (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA_CURRENT)) {
+        *cuda_mem_type = CU_MEMORYTYPE_DEVICE;
+    } else {
+        status = UCS_ERR_INVALID_ADDR;
+    }
+
+mem_release:
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(cuMemRelease(alloc_handle));
+ret:
+#endif
+    return status;
+}
+
+static void
+uct_cuda_copy_set_sync_memops(const void *address, unsigned is_vmm_allocation)
+{
+    unsigned value = 1;
+
+#if HAVE_CUDA_FABRIC
+    if (is_vmm_allocation) {
+        UCT_CUDADRV_FUNC_LOG_WARN(cuCtxSetFlags(CU_CTX_SYNC_MEMOPS));
+    } else
+#endif
+    {
+        UCT_CUDADRV_FUNC_LOG_WARN(cuPointerSetAttribute(&value,
+                    CU_POINTER_ATTRIBUTE_SYNC_MEMOPS, (CUdeviceptr)address));
+    }
+}
+
 static ucs_status_t
 uct_cuda_copy_md_query_attributes(uct_cuda_copy_md_t *md, const void *address,
                                   size_t length, ucs_memory_info_t *mem_info)
 {
 #define UCT_CUDA_MEM_QUERY_NUM_ATTRS 4
-    CUmemorytype cuda_mem_mype = (CUmemorytype)0;
+    CUmemorytype cuda_mem_type = (CUmemorytype)0;
     uint32_t is_managed        = 0;
     CUdevice cuda_device       = -1;
     CUcontext cuda_mem_ctx     = NULL;
+    unsigned is_vmm_allocation = 0;
```

Comment:
maybe replace local carbs is_vmm / is_managed by a single var of the type uct_cuda_alloc_type_t?

Reply:
We need is_managed to pass to pointer attribute query API so we wouldn't be getting rid of another variable anyway.